### DISCUSSION TITLE
How to use nixpkgs directly from lua?

### DISCUSSION THREAD

**Discussion by starptr** [nix-community/kickstart-nix.nvim#52] [Q&A] [answered]
I'm trying to use a binary, e.g. `pkgs.hello`, from lua. Although `vim.system` can use a binary that's added in `extraPackages` in the `neovim-overlay.nix`, I believe it is simply looking up the binary from path, which is a bit impure. Is it possible to explicitly call a package from nixpkgs?

**Comment by mrcjkb** [comment]
Calling a package from nixpkgs would be impure, too.
I'm guessing you mean you want Neovim to only see packages that are included in `extraPackages`?

You could build a wrapper that unsets `PATH` first. I don't think you'd really want to do that though. The point of Nix is not to isolate the environment of a package at runtime, it's to have input-reproducibility when building. 

I prefer to use devShells to manage external dependencies. For example, some language servers can slow you down if you only want to quickly look at something. By having them in devShells, I can easily disable them without having to change my Neovim config.

**Reply by starptr** [reply]
Not exactly. As an example of a wrapped nix package (i.e. not neovim), you can write bash scripts inline in nix like e.g. `${pkgs.ripgrep}/bin/rg foo`. So, you can package bash scripts in nix which have runtime dependencies on binaries both inside and outside of the nix store. Whichever is more appropriate definitely depends on a case-by-case basis. I think the lua init scripts in kickstart-nix.nvim are similar; sometimes you want them to have runtime dependencies that are *not* in the nix store like in your example so that the caller of the neovim binary can swap language servers at runtime. But there are scenarios where the lua init script depends on something that shouldn't need to change (i.e. the caller of the neovim binary should not be able to accidentally change/remove the runtime dependency), and for those situations, being able to refer directly to the nix store is nice.

As a concrete example, I'm invoking an executable that tells neovim what the system theme is. The caller of my neovim distribution should not be able to "dependency inject" this executable. If I can directly refer to the binary in the nix store from my lua script, then it is more durable than if I'm just referring to the binary by name. In other words, how can I do the `${pkgs.ripgrep}/bin/rg` trick (like in a bash script), but in a lua init script instead?

**Reply by mrcjkb** [reply, marked as answer]
You can do that via plugin configs.
Plugins don't have to be a derivation, they can also be [an attrset](https://github.com/nix-community/kickstart-nix.nvim/blob/2fad4f4a4103d805b91bb4118629cad8a988b6ec/nix/mkNeovim.nix#L42).

**Reply by starptr** [reply]
Thanks. For posterity: you can write vimscript in the plugin attrset's `config` property. This property ends up [here](https://github.com/hsjobeki/nixpkgs/blob/046fb425c8d94a917558603010e18e8e52faf063/pkgs/applications/editors/neovim/wrapper.nix#L80), which is inlined [here](https://github.com/hsjobeki/nixpkgs/blob/046fb425c8d94a917558603010e18e8e52faf063/pkgs/applications/editors/neovim/wrapper.nix#L87), which ends up [here](https://github.com/hsjobeki/nixpkgs/blob/046fb425c8d94a917558603010e18e8e52faf063/pkgs/applications/editors/neovim/wrapper.nix#L103) as an IFD file containing vimscript, which `init.lua` will call at the end.

---

Not sure if this is related to the original question, but [what does the runtime property do](https://github.com/nix-community/kickstart-nix.nvim/blob/2fad4f4a4103d805b91bb4118629cad8a988b6ec/nix/mkNeovim.nix#L49)? I couldn't find mention of it in nixpkgs.

**Reply by mrcjkb** [reply]
You don't have to use Vimscript, you [can also use Lua](https://github.com/hsjobeki/nixpkgs/blob/046fb425c8d94a917558603010e18e8e52faf063/pkgs/applications/editors/neovim/wrapper.nix#L96).

`wrapNeovimUnstable` is (as the name suggests) unstable and changes quite often (The reason I see it as safe to use is because I can pin it in my flake).

I'm not sure what the `runtime` attr is/was for. It may be an old attribute that has been removed by now.

**Reply by starptr** [reply]
Oh yes that's a good point. I'm a little more hesitant to use `p.plugin.passthru.initLua` atm because `p.config` is mentioned in [the util function to normalize the plugin attrset](https://github.com/hsjobeki/nixpkgs/blob/046fb425c8d94a917558603010e18e8e52faf063/pkgs/applications/editors/neovim/utils.nix#L26), whereas the `initLua` attribute isn't mentioned anywhere else in nixpkgs (although I did see the effort to address this  in NixOS/nixpkgs#172538). Also, I didn't realize `wrapNeovimUnstable` meant that the function itself has an unstable interface 😅, so that's good to keep in mind if I update the flake inputs.
